feat: use `semver` to match required version (#6066)
rust-lang/rustfmt: Configurations.md

specific version of rustfmt is used in your CI, use this option.

- **Default value**: `CARGO_PKG_VERSION`

> - **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/) and following the following [`semver::Version`](https://docs.rs/semver/latest/semver/struct.Version.html#total-ordering) specification.

Reviewer: Let's drop the note about `semver::Version`. That's an internal implementation detail that I don't think we need to mention in the docs.

```suggestion
- **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/).
```

  - Multiple values can be used, they should be separated by a comma, e.g.: `1.0.0, 1.1.0`. The comparison is done using the `&&` operator, so all versions from a given range must be satisfied.
  - By default, values without comparison operators are treated as `^` (caret) version requirements.
  - If you want to specify an exact version, use the operator `=`, e.g.: `=1.0.0`.

>   - If you want to specify a range, use the operator `, `, e.g.: `1.0.0, <2.0.0`. This is the same as `^1.0.0, <2.0.0` or `^1.0.0`.

Reviewer: These explanations are good, but I think it would be more valuable for users to see exactly how to specify these in their rustfmt.toml files. For consistency with the other options let's move this entire bulleted list below the `**Stable**` bullet. Having put a little more thought into this I think it might even be more useful to define subsections for the given semver targeting instead of bullets. Something like:

````markdown
### Specify a version range
```toml
required_version="1.0.0, <2.0.0"
```
````

>  By default, values without comparison operators are treated as `^` (caret) version requirements.

Two notes:
1. Users who aren't that familiar with semver might be confused by what the `^` means. Can we change this to something like: `Match on new minor or patch versions`.
2. I want to think a little more on the actual semantics of this. Right now I'm leaning towards keeping the existing behaviour where `required_version=1.0.0` means an exact match for 1.0.0. I know this means we'll need to workaround how the `semver` crate wants to parse the `VersionReq`. We can brainstorm different ideas for how to do that. let me know what your thoughts are on this.

Author: > I want to think a little more on the actual semantics of this. Right now I'm leaning towards keeping the existing behaviour where required_version=1.0.0 means an exact match for 1.0.0. I know this means we'll need to workaround how the semver crate wants to parse the VersionReq. We can brainstorm different ideas for how to do that. let me know what your thoughts are on this.

Agree with you! While I did use the standard `semver` behavior to avoid complexity, I wrote a draft matching the exact version: https://github.com/rust-lang/rustfmt/pull/6066/commits/73e75adc63e337c048cfdac380710b7dd1939ba7
Now that we split `check_semver_version`, it'll be easier to define this behavior.

Author: Changed the docs + the behavior. Let me know if it requires additional tweaks

specific version of rustfmt is used in your CI, use this option.

- **Default value**: `CARGO_PKG_VERSION`
- **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/). Multiple values can be used, they should be separeted by a comma, e.g.: `1.0.0, 1.1.0`. By default, values without comparison operators are treated as `^` (caret) version requirements. If you want to specify an exact version, use the operator `=`, e.g.: `=1.0.0`.
- **Stable**: No (tracking issue: [#3386](https://github.com/rust-lang/rustfmt/issues/3386))

> 

Reviewer: I think the first sentence is good and can stay here for the  `Possible values`, but I think we should move the rest of the explanation to a small paragraph below. Maybe even a bulleted list showing users how to target exact, major, minor, and patch versions. Oh and maybe a range of versions too like you mentioned here https://github.com/rust-lang/rustfmt/issues/6063#issue-2125707100.

specific version of rustfmt is used in your CI, use this option.

- **Default value**: `CARGO_PKG_VERSION`
- **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/).
- **Stable**: No (tracking issue: [#3386](https://github.com/rust-lang/rustfmt/issues/3386))

> #### Match on exact version

Reviewer: nit: all other titles end with a `:`. For consistency let's do the same here.

```suggestion
#### Match on exact version:
```

specific version of rustfmt is used in your CI, use this option.

- **Default value**: `CARGO_PKG_VERSION`
- **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/).
- **Stable**: No (tracking issue: [#3386](https://github.com/rust-lang/rustfmt/issues/3386))

> #### `CARGO_PKG_VERSION` (default):
> 
> Match on exact version

Reviewer: Not sure if we need to specify ``#### `CARGO_PKG_VERSION` (default):`` Probably nicer to stay consistent with what you've done for the other examples, which I think are very clear and helpful!

```suggestion
#### Match on exact version
```

```toml
required_version="1.0.0"
```

#### Higher or equal to:

```toml
required_version=">=1.0.0"
```

#### Lower or equal to:

```toml
required_version="<=1.0.0"
```

#### New minor or patch versions:

```toml
required_version="^1.0.0"
```

#### New patch versions:

```toml
required_version="~1.0.0"
```

#### Wildcard:

```toml
required_version="*" # matches any version.
required_version="1.*" # matches any version with the same major version
required_version="1.0.*" # matches any version with the same major and minor version
```

#### Multiple versions to match:

> Any of the semver operators can be combined, being split with commas. The comparison is done using `&&` operator.
> 
> Can be used with any operator except by `*`, as any comparator will override the `*` operator. If you use `*` with other comparators, the comparison will fail.

Author: ```suggestion
Except by `*`, can be used by any of the semver operators can be combined, being split with commas. The comparison is done using `&&` operator.

`*` has range restrictions as any comparator will override the wildcard operator. When `*` is used within a range, the comparison always fails.
```

- **Possible values**: `semver` compliant values, such as defined on [semver.org](https://semver.org/).
- **Stable**: No (tracking issue: [#3386](https://github.com/rust-lang/rustfmt/issues/3386))

#### Match on exact version:

```toml
required_version="1.0.0"
```

#### Higher or equal to:

```toml
required_version=">=1.0.0"
```

#### Lower or equal to:

```toml
required_version="<=1.0.0"
```

#### New minor or patch versions:

```toml
required_version="^1.0.0"
```

#### New patch versions:

```toml
required_version="~1.0.0"
```

#### Wildcard:

```toml
required_version="*" # matches any version.
required_version="1.*" # matches any version with the same major version
required_version="1.0.*" # matches any version with the same major and minor version
```

#### Multiple versions to match:

Except by `*`, any of the semver operators can be combined, being split with commas. The comparison is done using `&&` operator.

`*` can't be used with other comparators.

Since `*` has range restrictions, any comparator will override the wildcard operator. When `*` is used alone, the comparison always fails, as that's an invalid config. `*, <1.0.0` for example is invalid.

> Version requirements can't contradict themselves, otherwise they'll always fail. Some examples are `"1.*, >2.0.0"`, `1.0.*, >2.0.0` and `<1.5.0, >1.10.*`, because both requirements can't be true at the same time.

Reviewer: I think we could reword this slightly to hopefully make it clearer:

```
A comma separated list of version requirements.
The match succeeds when the current rustfmt version matches all version requirements.

The one notable exception is that a wildcard matching any version cannot be used in the list.
For example, `*, <1.0.0` will always fail.

Additionally, the version match will always fail if any of the version requirements contradict themselves.
Some examples of contradictory requirements are `1.*, >2.0.0`, `1.0.*, >2.0.0` and `<1.5.0, >1.10.*`.
```

Reviewer: @ologbonowiwi Do you have any thoughts on this potential rewording?

Author: Hi @ytmimi, sorry for the late response!

Agree with you. Will commit the change

#### Match on exact version:

```toml
required_version="1.0.0"
```

#### Higher or equal to:

```toml
required_version=">=1.0.0"
```

#### Lower or equal to:

```toml
required_version="<=1.0.0"
```

#### New minor or patch versions:

```toml
required_version="^1.0.0"
```

#### New patch versions:

```toml
required_version="~1.0.0"
```

#### Wildcard:

```toml
required_version="*" # matches any version.
required_version="1.*" # matches any version with the same major version
required_version="1.0.*" # matches any version with the same major and minor version
```

#### Multiple versions to match:

Except by `*`, any of the semver operators can be combined, being split with commas. The comparison is done using `&&` operator.

`*` can't be used with other comparators.

Since `*` has range restrictions, any comparator will override the wildcard operator. When `*` is used alone, the comparison always fails, as that's an invalid config. `*, <1.0.0` for example is invalid.

Versions can't contradict themselves, otherwise they'll always fail. Some examples are `"1.*, <2.0.0"` and `1.0.*, <2.0.0`, since both versions can't be true at the same time.

> By now, `&&` (using comma `,`) is the only way to declare `required_version`. `||` Operator is not valid. Valid examples are `1.0.0, 1.1.0` and `1.0.0, 1.1.0, 1.1.*` but not `1.0.0 || 1.1.0` or `1.0.0, 2.0.0`.

Author: Supporting `||` matters to us, or can we keep only the comma operator?

Author: Sure, was too into this, mb. 

So, the `semver` crate works so that you can send multiple versions, which should be compatible. It is like the use of an "and" operator. So your version can be something like `^1.0.0, 1.0.*`. But can't be `^1.0.0, ^2.0.0`.

My question is whether it's important to us to support multiple versions, which are not necessarily always true. 

Are we okay with multiple versions translating from `^1.0.0, ^2.0.0` being (matches version a AND version b)? or do we want this to be "matches version a OR version b".

Let me know if it's clearer now.

Reviewer: That's clearer, thanks. Do other tools with semver functionality support `OR` with `||`? Does the `semver` crate support that functionality? If not then I don't think it's necessary. That said, I don't want to increase the scope of the current PR so I'd still want to hold off on `||` until a follow up PR.

Author: I don't think so unless they're implementing it manually. 

Agree with you that we can go as it is and implement a follow-up PR.

Reviewer: Given that we're not going to support `OR` logic with `||` I'd prefer it if we removed that language from the docs that you wrote.

Also I'm having a hard time understanding why the following example is contradictory:

> Versions can't contradict themselves, otherwise they'll always fail. Some examples are `"1.*, <2.0.0"` and `1.0.*, <2.0.0`, since both versions can't be true at the same time.

Why can't both be true? Isn't the first example communicating that we'll accept any `v1` and the second is saying we want any `v1.0.*`? The `<2.0.0` feels redundant, but I don't think it's contradictory. Maybe an example like `<1.5.0, >1.10.*` would make more sense for something that's contradictory?

Author: Yeah actually the example is wrong; what I meant is `1.*, >2.0.0`

Author: Fixed the docs and included your example.

See also [`max_width`](#max_width), [`use_small_heuristics`](#use_small_heuristics), and [`struct_lit_single_line`](#struct_lit_single_line)

## `style_edition`

Controls the edition of the [Rust Style Guide] to use for formatting ([RFC 3338])	

- **Default value**: `"2015"`	
- **Possible values**: `"2015"`, `"2018"`, `"2021"`, `"2024"` (unstable variant)	
- **Stable**: No	

[Rust Style Guide]: https://doc.rust-lang.org/nightly/style-guide/	
[RFC 3338]: https://rust-lang.github.io/rfcs/3338-style-evolution.html	

> 

Reviewer: Please revert this change. I don't want to make any changes to other config documentation in this PR.

Author: fixed. it was caused by a merge conflict

See also [`max_width`](#max_width) and [`use_small_heuristics`](#use_small_heuristics)

## `style_edition`

Controls the edition of the [Rust Style Guide] to use for formatting ([RFC 3338])

- **Default value**: `"2015"`
- **Possible values**: `"2015"`, `"2018"`, `"2021"`, `"2024"` (unstable variant)
- **Stable**: No

[Rust Style Guide]: https://doc.rust-lang.org/nightly/style-guide/
[RFC 3338]: https://rust-lang.github.io/rfcs/3338-style-evolution.html

> 

Reviewer: I think you rebased this incorrectly. This should not have been removed.

See also [`max_width`](#max_width) and [`use_small_heuristics`](#use_small_heuristics)

## `style_edition`

Controls the edition of the [Rust Style Guide] to use for formatting ([RFC 3338])

- **Default value**: `"2015"`
- **Possible values**: `"2015"`, `"2018"`, `"2021"`, `"2024"` (unstable variant)
- **Stable**: No

[Rust Style Guide]: https://doc.rust-lang.org/nightly/style-guide/
[RFC 3338]: https://rust-lang.github.io/rfcs/3338-style-evolution.html

> 

Reviewer: Please revert this change. I don't want to make any changes to other config documentation in this PR.

Author: [this change](https://github.com/rust-lang/rustfmt/pull/6066/files/fede8d0de367185aa403780b595ca76d10282a52#r1795880128) you mentioned here should actually be here. will fix it.

## `version`

Which version of the formatting rules to use. `Version::One` is backwards-compatible
with Rustfmt 1.0. Other versions are only backwards compatible within a major

> version number.

Reviewer: This should not have been removed
